### ISSUE TITLE
Marking a BlockingCollection with CompleteAdding() later causes System.OperationCanceledException to be thrown

### ISSUE THREAD

**Issue by mayka-mack** [dotnet/runtime#97946]
### Description

Starting with .NET 8.0, if you mark a `BlockingCollection` with `CompleteAdding()`, it causes subsequent execution to fail under certain conditions. This may possibly occur if the collection has no remaining items.

### Reproduction Steps

The below code snippet will error on the line after `queue.CompleteAdding();`.
```
BlockingCollection<string> queue = new();
Task task = Task.Run(() =>
{
    foreach (string next in queue.GetConsumingEnumerable())
    {
        Console.WriteLine(next);
    }
});

queue.Add("Item One");
Thread.Sleep(100); // pause will trigger the error
queue.CompleteAdding();
Console.Read(); // code errors here
```

### Expected behavior

For execution to continue after the `BlockingCollection` is marked with `CompleteAdding()`, with the program only erroring if more items are attempted to be added to the `BlockingCollection`.

### Actual behavior

The next line of code after `queue.CompleteAdding()` causes a `System.OperationCanceledException` to be thrown within `System.Threading.SemaphoreSlim.Wait(int, CancellationToken)`.

```
System.OperationCanceledException
  HResult=0x8013153B
  Message=The operation was canceled.
  Source=System.Private.CoreLib
  StackTrace:
   at System.Threading.SemaphoreSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/SemaphoreSlim.cs:line 390
   at System.Collections.Concurrent.BlockingCollection`1.TryTakeWithNoTimeValidation(T& item, Int32 millisecondsTimeout, CancellationToken cancellationToken, CancellationTokenSource combinedTokenSource)
```

### Regression?

This behavior does not occur within .NET 7.0 or earlier.

### Configuration

| Config       | Value         |
|--------------|---------------|
| .NET Version | 8.0.1         |
| OS Version   | Windows 10    |
| Architecture | x64           |

Note: This error is not platform-specific, as it also reproduced on WSL.

### Other information

Behavior first noticed in https://github.com/serilog/serilog-sinks-async/issues/91

**Comment by stephentoub** [comment]
I can't repro this. Is the above repro code 100% deterministic in failing for you?

**Comment by mayka-mack** [comment]
> I can't repro this. Is the above repro code 100% deterministic in failing for you?

Yes, the above repro code 100% deterministic in failing for me. I have done some more digging, and it looks like the error is throwing specifically when using the Visual Studio debugger. When I ran the same code through the debugger in LinqPAD and Rider it did not throw the error, and .NET 7.0 and earlier did not throw an error in the Visual Studio debugger.

**Comment by mayka-mack** [comment]
I've upgraded to Visual Studio 17.9.0 and am no longer seeing this behavior when using the debugger on Windows. I'm still seeing an exception thrown in the Debug output from this code when running under WSL in VS, though these exceptions do not appear when directly running the executable under WSL, only through Visual Studio's debugger.
```
Item One
Exception thrown: 'System.OperationCanceledException' in System.Private.CoreLib.dll
Exception thrown: 'System.OperationCanceledException' in System.Private.CoreLib.dll
```
I'm not sure if this is VS-related only with no ties to .NET 8.0, or if they are intertwined somehow. Should this issue be closed, or would it require more investigation?

**Comment by stephentoub** [comment]
> Should this issue be closed

I think so. It seems like you might just being seeing information about first-chance exceptions, exceptions that end up getting thrown but then caught rather than allowed to propagate out to consuming code.

**Comment by mayka-mack** [comment]
> > Should this issue be closed
> 
> I think so. It seems like you might just being seeing information about first-chance exceptions, exceptions that end up getting thrown but then caught rather than allowed to propagate out to consuming code.

I believe you are correct. I think the only issue in this case was Visual Studio breaking on these first-chance exceptions in the debugger, which seems to have hopefully been corrected in 17.9.
